Add Fortran elastic-tube-1d solvers (precice Fortran module)

YonatanGM opened the pull request:
This PR adds Fortran solvers for the elastic-tube-1d case using the fortran module (precice.f90). I also opened PR #606 that doesn't rely on this module.

The `run.sh` script downloads precice.f90 from the Fortran module repository to get the latest version and the file gets compiled with the solvers source files using CMake. Since the module is just a wrapper around the precicef functions in the library, which are updated with preCICE release, I imagine it will stay compatible with the precice version used in the solvers (as long as the function names stay consistent)

LAPACK is needed for the fluid solver (same as the C++ case). Just run ./run.sh to build and execute via CMake. The plotting scripts (plot-diameter.sh and plot-all.sh) are updated to show results for the Fortran case. The precommit yaml didnt have hooks for Fortran files, so I maunally formatted them using fprettify.  

Results look consistent with the other solvers. I checked the VTK output, and the numbers match up to 5 decimal places on the last timestep. Also tested running the fortran case against the other solves and that seems to work fine too. 

![tutorials-elastic-tube-1d-all](https://github.com/user-attachments/assets/926fd428-6d88-4556-9600-6a1881e8ed29)
![Figure-fortran](https://github.com/user-attachments/assets/aaaa1ba4-3fe1-4ff7-8214-68e4690ff46d)
![Figure-python](https://github.com/user-attachments/assets/bb88b87d-ae04-41e4-9893-13a46e98dfc5)


Checklist:

- [x] I added a summary of any user-facing changes (compared to the last release) in the `changelog-entries/<PRnumber>.md`.
- [ ] I will remember to squash-and-merge, providing a useful summary of the changes of this PR.
- [x] Update the image used in readme showing diameter and pressure of running all solvers against cpp to inlcude fortran case

YonatanGM committed:
Create initial folders and files

YonatanGM committed:
Add scripts, test with minimal solvers

YonatanGM committed:
Implementation wip

YonatanGM committed:
Add output folder

YonatanGM committed:
Implement vtk writing, Add missing write_data call in fluid solver

YonatanGM committed:
Set domainSize to 100

YonatanGM committed:
Use ES format to output in scientific notation

YonatanGM committed:
Refactor: clean code, use snake_case for subroutines, apply dp for consistent precision

YonatanGM committed:
Remove unused var

YonatanGM committed:
apply dp for consistent precision

YonatanGM committed:
Remove debug prints

YonatanGM committed:
Remove comment

YonatanGM committed:
Include Fortran watchpoint in plot-all script

YonatanGM committed:
Improve print statement formatting

YonatanGM committed:
Remove debug compiler flags and clean up

YonatanGM committed:
Set Fortran standard and enable warnings in debug builds

YonatanGM committed:
Remove unused variable

YonatanGM committed:
Set character lengths to 50

YonatanGM committed:
Add preCICE Fortran module file (precice.f90)

YonatanGM committed:
Add preCICE Fortran module file (precice.f90)

YonatanGM committed:
Add Fortran module and update precicef_ calls with length args

YonatanGM committed:
Add plotting for fluid fortran diameter

YonatanGM committed:
Delete file (now fetched dynamically)

YonatanGM committed:
Get precice.f90 from the fortran module repo in run script

MakisH commented on elastic-tube-1d/fluid-fortran/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+cmake_minimum_required(VERSION 3.10)
+project(ElasticTube LANGUAGES Fortran C)
```
Just curious (I don't know): why is the `C` required here?

YonatanGM replied:
Without the C, I get a compilation error in the find_package(precice...) call, which seems to use the FindThread cmake module and this module requires C or C++ to set up threading support. SO we need to specify C or CXX.

MakisH commented on elastic-tube-1d/fluid-fortran/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+cmake_minimum_required(VERSION 3.10)
+project(ElasticTube LANGUAGES Fortran C)
+
+set(CMAKE_Fortran_STANDARD 2003)
```
Does the variable `CMAKE_Fortran_STANDARD` exist or do anything?
From [this CMake issue](https://gitlab.kitware.com/cmake/cmake/-/issues/25006), I understand that setting Fortran standards is not intrinsically supported by CMake at the moment.

If removed, don't forget to also apply this (and any further changes) to the `solid-fortran/CMakeLists.txt` and in the other PR.

YonatanGM replied:
you're right, CMake doesn't seem to support Fortran standards.  I just removed it in both PRs. 
I think we could still specify it like so: add_compile_options($<$<COMPILE_LANG_AND_ID:Fortran,GNU>:-std=f2003>)

MakisH replied:
I only have a very limited understanding of Fortran standards, but I have heard that they work a bit differently than the C++ standards. Let's not set anything and leave the defaults (i.e., as you already have it now).

MakisH commented on elastic-tube-1d/fluid-fortran/run.sh:
```diff
@@ -0,0 +1,20 @@
+. ../../tools/log.sh
+exec > >(tee --append "$LOGFILE") 2>&1
+
+if [ ! -f src/precice.f90 ]; then
+  echo "Fetching precice.f90 (Module for Fortran bindings of preCICE)..."
+  curl -o src/precice.f90 https://raw.githubusercontent.com/precice/fortran-module/master/precice.f90
+fi
```
I think this step is fine, given that it first checks if the file exists, so one could skip it (we need to document this in a `fluid-fortran/README.md`, though).

However, what was your motivation for getting the source file, instead of linking to the module, as we do in the [example](https://github.com/precice/fortran-module/blob/develop/examples/solverdummy/README.md)? It is definitely easier for the user, I am just wondering if there are further reasons.

YonatanGM replied:
Getting precice.f90 and using CMake to compile it with the solvers felt simpler. The [example](https://github.com/precice/fortran-module/tree/develop) uses two Makefiles, one to generate the .mod file and another to build the solver. We could still streamline this in the run script (by just adding the Make commands), but I preferred CMake to match the C++ case. Also, I had in mind that, CMake was more flexible because it can generate platform specific files like VS code solutions and detect compilers (the makefiles are hardcoded to use GNU fortran i think). Also don't know if makefiles are supported on windows. Let me know what you think!

MakisH replied:
I see your point. Directly building with the same stack would be more robust, and pulling from the internet should be possible on all the relevant systems. Let's keep it like this.

MakisH commented on elastic-tube-1d/fluid-fortran/src/FluidComputeSolution.f90:
```diff
@@ -0,0 +1,182 @@
+        L = 10.0
+        dx = L/kappa  !1.0 / (N * kappa);
+
+        ! result variable
```
Since this `info` is now (in the Fortran version) far away from where it is used, we need a more descriptive comment.

I would argue that we need a better name in all versions, but let's keep it as-is now, for consistency and to keep things simple.

YonatanGM replied:
I have updated the comment here

MakisH commented on elastic-tube-1d/fluid-fortran/src/FluidComputeSolution.f90:
```diff
@@ -0,0 +1,182 @@
+            ! Initialize residual vector
+            Res = 0.0
+
+            ! Compute residuals
+            do i = 2, N  ! Adjusted for 1-based indexing
```
I have not checked all the indices here, but I looked at several, and it looks consistent with the C++ version. The bounds make sense.

YonatanGM replied:
I double checked them too and they seem correct.

MakisH commented on elastic-tube-1d/fluid-fortran/src/FluidComputeSolution.f90:
```diff
@@ -0,0 +1,182 @@
+        real(dp), allocatable :: lhs(:, :)
+        integer, allocatable :: ipiv(:)
+
+        nlhs = 2*N + 2
```
There seems to be a mix-up between `n` and `N`. Let's use one of the two forms, as at least it helps text finding. The C++ version uses `N`.

Same with `res`/`Res`, `lhs`/`LHS` and maybe more.

(Remember: in Fortran, these two are the same variable)

YonatanGM replied:
I fixed this!

MakisH reviewed (commented):
Hi @YonatanGM,

wearing my preCICE hat here :cowboy_hat_face: (it is a bit unfortunate that I am also supervising your SSE project, maybe @fsimonis can control my conflict of interest).

With a first look, I think we could keep both contributions, since we support both ways of using preCICE via Fortran. Let's start with this one, and at the same time port any changes to the other one as well (should not be too many). Not sure yet how we could reduce duplication, but I have the impression the simplest would be to keep both, in different directories (`fluid-fortran` and `fluid-fortran-module`).

I had a first look and I ran the code, mixing languages to check for potential incompatibilities. I arrive at the same results.

![tutorials-elastic-tube-1d-all](https://github.com/user-attachments/assets/a8ff010d-9f53-4ac8-b2f4-e82157c91f9b)

In terms of the solver source code, I have only reviewed the `FluidComputeSolution.f90` so far, as well as the shell scripts and `CMakeLists.txt`. I will check the rest of the Fortran files during the day.

MakisH commented on elastic-tube-1d/fluid-fortran/src/FluidSolver.f90:
```diff
@@ -0,0 +1,194 @@
+  character(LEN=50) :: configFileName
+  character(LEN=50)  :: solverName
+  character(LEN=50)  :: meshName, pressureName, crossSectionLengthName
+  character(LEN=256) :: outputFilePrefix
```
I agree that 256 would probably be a better value than 50 everywhere (while we need to eventually drop this need altogether), but what led you give a larger value here? Any particular issues?

YonatanGM replied:
Actually i meant to use 256 for the full filename. The prefix is hard coded like this: 
 ```fortran
outputFilePrefix = './output/out_fluid'  
```
So i changed it to 50 like the others.

MakisH commented on elastic-tube-1d/fluid-fortran/src/FluidSolver.f90:
```diff
@@ -0,0 +1,194 @@
+  t_shift = 0.0_dp
+  p0 = 0.0_dp
+  vel_in_0 = u0 + ampl * sin(frequency * (t_shift) * pi)
+
```
The C++ version has some comments on what these paramteres are. Since these are parameters someone might need to change at some point, let's port the comments here as well.

MakisH commented on elastic-tube-1d/fluid-fortran/src/utilities.f90:
```diff
@@ -0,0 +1,74 @@
+    integer :: ioUnit, i, ioStatus
+    character(LEN=256) :: filename
+
+    write (filename, '(A,"_",I0,".vtk")') trim(filenamePrefix), iteration
```
The `filenamePrefix` was already defined as 256 characters long, and `filename` is also defined as 256 characters long. So this could potentially fail, right?

YonatanGM replied:
Yes, this is wrong. The prefix is small and hard-coded right now, so I changed it to 50. For the full filename, I still use 256 (it's probably too much for this tutorial case, I think there's only 100 time steps / iterations but probably safe to keep it this way)

MakisH reviewed (commented):
Some of my comments were still somehow pending publishing. @YonatanGM let me know if everything is clear.

YonatanGM commented:
> Some of my comments were still somehow pending publishing. @YonatanGM let me know if everything is clear.

Hi @MakisH, Thanks for the comments!! I will address them later today or tomorrow.

YonatanGM committed:
Add descriptive comment for info, fix variable capitalizations

YonatanGM committed:
Use 50 chars for  outputFilePrefix

YonatanGM committed:
Add comments from cpp, capitalize pi

YonatanGM committed:
Remove Fortran standard

YonatanGM committed:
Added -module suffix to solver folders

YonatanGM committed:
Include fortran result in plot-all image

YonatanGM commented:
> Hi @YonatanGM,
> 
> wearing my preCICE hat here 🤠 (it is a bit unfortunate that I am also supervising your SSE project, maybe @fsimonis can control my conflict of interest).
> 
> With a first look, I think we could keep both contributions, since we support both ways of using preCICE via Fortran. Let's start with this one, and at the same time port any changes to the other one as well (should not be too many). Not sure yet how we could reduce duplication, but I have the impression the simplest would be to keep both, in different directories (`fluid-fortran` and `fluid-fortran-module`).
> 
> I had a first look and I ran the code, mixing languages to check for potential incompatibilities. I arrive at the same results.
> 
> ![tutorials-elastic-tube-1d-all](https://private-user-images.githubusercontent.com/4943683/403326739-a8ff010d-9f53-4ac8-b2f4-e82157c91f9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5NTc5ODUsIm5iZiI6MTczNzk1NzY4NSwicGF0aCI6Ii80OTQzNjgzLzQwMzMyNjczOS1hOGZmMDEwZC05ZjUzLTRhYzgtYjJmNC1lODIxNTdjOTFmOWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjdUMDYwMTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWI0MGJiMjJkNDI5ZjNhZmQxYzA4NDFhOWQ1ZGU2Njg2ZTJhOGIzODBjOTA2NjgzZmJhNTE3ZWM2ZTA1NTkwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sTmBQFeseTco10UysukPiqgLKcmD-mWOLJmllosPn8g)
> 
> In terms of the solver source code, I have only reviewed the `FluidComputeSolution.f90` so far, as well as the shell scripts and `CMakeLists.txt`. I will check the rest of the Fortran files during the day.

Hi @MakisH, I just addressed your suggestions! I renamed the folder to `fluid/solid-fortran-module`. PR #606 uses `fluid/solid-fortran`.

YonatanGM committed:
Deleted folders

YonatanGM commented:
Hi @MakisH, I addressed your suggestions! I renamed the folder to `fluid/solid-fortran-module`. PR #606 uses `fluid/solid-fortran`.
One problem is the `plot-all.sh` and `plot-diameter.sh` scripts in this PR still reference the `fluid-fortran`folder (which is not there anymore), so if you try to run those scripts they will fail. But once both PRs are merged there should no issue. The results from both PRs are the same, so I though it makes sense to just plot one of them.

YonatanGM commented:
@MakisH I noticed there are no precice finalize calls in the C++ and python versions. Could that be problematic?

MakisH commented on elastic-tube-1d/plot-all.sh:
> One problem is the `plot-all.sh` and `plot-diameter.sh` scripts in this PR still reference the `fluid-fortran`folder (which is not there anymore), so if you try to run those scripts they will fail. But once both PRs are merged there should no issue. The results from both PRs are the same, so I though it makes sense to just plot one of them.

@YonatanGM I would suggest to treat the `solid-fortran` and the `solid-fortran-module` as two distinct variants, and plot both. Since this case is also used as a test, it makes sense to run both and compare the results. They check different components.

This means: add a `solid-fortran-module` entry in this file and in the `plot-diameter.sh`.

MakisH commented on elastic-tube-1d/fluid-fortran-module/output/.keepme:
Is this actually needed?
This is only for the preCICE export files, right? In recent versions, the directory does not need to pre-exist.

YonatanGM replied:
@MakisH  hmm, I'm not sure actually (the cpp and python versions have that folder with the .keepme in it). 
I just removed the .keepme and pushed it but git doesnt track empty folders it seems so the output folder is now gone. 
Should we create the folder when writing the vtk outputs maybe? or in the run script ?

MakisH replied:
You are right, I had the impression this was the output of preCICE, but we are actually not writing any export files in this tutorial. I checked again, and while the code works, it indeed does not produce files.

Let's revert the change and add the `.keepme`, but please add a comment in the file on why this is needed (pointing to the `FluidSolver.f90`).

MakisH replied:
Actually, creating the folder in the run script would be even better.

YonatanGM replied:
Just committed this

MakisH reviewed (commented):
Thank you very much for the prompt and very clean updates. Looks already very good! Just a couple of tiny points, mostly replying to your questions.

MakisH commented:
> @MakisH I noticed there are no precice finalize calls in the C++ and python versions. Could that be problematic?

Good observation, thanks for asking!

In the C++ API, the `finalize()` is called in the destructor of the `Participant` class, so it is called automatically when the program exits (i.e., when the object goes out of scope). This is a good practice, which has the strange name "Resource Acquisition Is Initialization (RAII)". I assume the Python bindings implement the same.

In C and Fortran, the whole paradigm is procedural, so we need to call this explicitly.

tl;dr: All good.

MakisH commented:
Also: Don't forget the changelog entry (in both PRs).

YonatanGM committed:
Include fluid-fortran-module results in plots

YonatanGM committed:
Update plot-all image

YonatanGM committed:
Fix backslash

MakisH commented:
The code works. Still pending:

- [x] ~~removing the `output/.keepme` file and the `output/` directory (I tested it, and it works without)~~ create the directory in the run script.
- [x] adding a changelog file `changelog-entries/607.md`

(putting my SSE hat on: these are trivial, but still have some learning significance)

YonatanGM committed:
Deleted file

YonatanGM committed:
Add changelog entry

YonatanGM commented:
@MakisH With both directory and file removed, it didnt work for me.
Getting errors like this:  Error: Unable to open file ./output/out_fluid_99.vtk
I think we have to keep the folder or create it first if it doesnt exist when writing the vtk files. What do you think?

YonatanGM committed:
Create output directory in run script

MakisH reviewed (approved):
Thank you for the very clean contribution and the very prompt and clear answers, @YonatanGM! I am merging this and I will merge #606 next.

MakisH commented:
After merging, I noticed one small thing that I would have done differently and slipped my review: the `precice.f90` is downloaded into `src/`. I would normally put it under `thirdparty/` (or even get the repository as a git module), to keep it clearly separated.

If you want to contribute this as well, that would be great, but I can also do it later.
